File: Unidad_2/ej7.py
# Obtenga el negativo de una imagen (grayscale y truecolor) y observe como se
# modifica su histograma. Hágalo aplicando la fórmula y con el comando
# imcomplement.

import matplotlib.pyplot as plt
from skimage import util
from skimage import img_as_float
from skimage.io import imread
import tkinter as tk
from tkinter import filedialog


def plot_img_hist(img, num_plot, title):
    """
        Plots de imagen e histograma
    """
    # Imagen
    plt.subplot(num_plot)
    plt.imshow(img, cmap='gray')
    plt.title(title)
    # Histograma
    plt.subplot(num_plot + 4)
    plt.hist(img_as_float(img).ravel(), bins=256)
    plt.xlim(0, 1)
    
def file_read():
    file_path = filedialog.askopenfilename()
    return file_path

try:
    # Busco imagen y obtengo su ruta
    root = tk.Tk()
    root.withdraw()
    img1 = imread(file_read()) # Cara
    img2 = imread(file_read()) # Tigre
    
    plot_img_hist(img1, 241, 'Cara')
    plot_img_hist(img2, 242, 'Tigre')
    plot_img_hist(util.invert(img1), 243, 'Cara invertida')
    plot_img_hist(util.invert(img2), 244, 'Tigre invertido')
    
    plt.show()
except:
    print('Cerraste la ventana!')

print('👋🏽')
root.destroy()

File: matlab/Unidad_2/ej7.m
% 7) Obtenga el negativo de una imagen (grayscale y truecolor) y observe
% como se modifica su histograma. Hágalo aplicando la fórmula y con el 
% comando imcomplement.

dirs = {}; % Creo una Cell para almacenar las dirs de las imagenes

for i=1:2
    [file, path] = uigetfile({'*.jpg';'*.JPG';'*.bmp';'*.tiff'});
    dir = strcat(path,file);
    info = imfinfo(dir);
    dirs{i} = imread(dir);
end

n = 8; % bits
subplot(2,3,1);
imshow(dirs{1});
title('Imagen True Color');

subplot(2,3,2);
imshow(imcomplement(dirs{1}));
title('Usando imcomplement');

subplot(2,3,3);
imshow((2^n - 1) - dirs{1});
title('Usando Bmax = (2^n - 1)');

subplot(2,3,4);
imshow(dirs{2});
title('Imagen Grayscale');

subplot(2,3,5);
imshow(imcomplement(dirs{2}));
title('Usando imcomplement');

subplot(2,3,6);
imshow((2^n - 1) - dirs{2});
title('Usando Bmax = (2^n - 1)');